packages/xforms-engine/src/lib/reactivity/text/createTextRange.ts

```diff
@@ -17,57 +20,84 @@ interface ChunksAndMedia {
 	mediaSources: MediaSources;
 }
 
+const generateChunk = (node: Node): TextChunkExpression<'string'> | null => {
+	if (isElementNode(node)) {
+		return TextChunkExpression.fromOutput(node);
+	}
+	if (isTextNode(node)) {
+		const formAttribute = node.parentElement!.getAttribute('form') as ResourceType;
+		if (isResourceType(formAttribute)) {
+			return TextChunkExpression.fromResource(node.data as JRResourceURLString, formAttribute);
+		}
+		return TextChunkExpression.fromLiteral(node.data);
+	}
+	return null;
+};
+
+const generateChunksForTranslation = (
+	textElement: Element
+): Array<TextChunkExpression<'string'>> => {
+	const chunks = [];
+	for (const child of textElement.childNodes) {
+		for (const grandchild of child.childNodes) {
+			const chunk = generateChunk(grandchild);
+			if (chunk) {
+				chunks.push(chunk);
+			}
+		}
+	}
+	return chunks;
+};
+
+const getChunkExpressions = <Role extends TextRole>(
+	context: EvaluationContext,
+	definition: TextRangeDefinition<Role>
+): ReadonlyArray<TextChunkExpression<'string'>> => {
+	if (definition.chunks[0]?.source !== 'translation') {
+		// only translations have 'nodes' chunks
+		return definition.chunks as Array<TextChunkExpression<'string'>>;
+	}
+	const itextId = context.evaluator.evaluateString(definition.chunks[0].toString()!, {
+		contextNode: context.contextNode,
+	});
+	const lang = context.getActiveLanguage();
+	const elem = definition.form.model.getItextElement(lang, itextId);
+	return elem ? generateChunksForTranslation(elem) : [];
+};
```

Comment:
This code has been moved here from the parsing step. By only chunking the translations for the active language we save a bunch of work for languages that may never be loaded.